### PR TITLE
Reworked chunking to support mega chunks

### PR DESCRIPTION
**Pull request by hagen-danswer**
## Description
- Changed how the chunker is instantiated and used throughout the indexing pipeline. 
- The chunker object now has more class functions and attributes. This was done to reduce passing around variables and improve clarity as  well as support creating a chunker object and a megachunker object
- Chunker object is no longer passed around and is, instead, grabbed right before usage (in indexing_pipeline.py) from a cache located in chunker.py. This was done to save memory when instantiating the largeish library "SentenceSplitter"
- 

## How Has This Been Tested?
- [ ] Run through the search eval pipeline to ensure identical performance

## Checklist:
- [ ] All of the automated tests pass
- [ ] All PR comments are addressed and marked resolved
- [ ] Docker images build and basic functionalities work
- [ ] Author has done a final read through of the PR right before merge

### CODE REVIEW

File: backend/danswer/configs/app_configs.py [hunk 1]
```diff
@@ -257,6 +257,12 @@
 # Slightly larger since the sentence aware split is a max cutoff so most minichunks will be under MINI_CHUNK_SIZE
 # tokens. But we need it to be at least as big as 1/4th chunk size to avoid having a tiny mini-chunk at the end
 MINI_CHUNK_SIZE = 150
+
+# Adding larger chunks for models that support it
+ENABLE_MEGA_CHUNK = os.environ.get("ENABLE_MEGA_CHUNK", "").lower() == "true"
```
Comment (rkuo-danswer) [inline backend/danswer/configs/app_configs.py]: Does it make sense to specifically define 2048 as a mega chunk, or would it be more flexible / future proof to simply allow the user to define the chunk size they want to use?

File: backend/danswer/indexing/indexing_pipeline.py [hunk 2]
```diff
@@ -263,15 +275,12 @@ def build_indexing_pipeline(
     embedder: IndexingEmbedder,
     document_index: DocumentIndex,
     db_session: Session,
-    chunker: Chunker | None = None,
     ignore_time_skip: bool = False,
 ) -> IndexingPipelineProtocol:
     """Builds a pipline which takes in a list (batch) of docs and indexes them."""
```
Comment (rkuo-danswer) [inline backend/danswer/indexing/indexing_pipeline.py]: typo "pipeline"

File: backend/danswer/indexing/chunker.py [hunk 3]
```diff
@@ -122,192 +72,247 @@ def _get_metadata_suffix_for_document_index(
...
-class Chunker:
-    @abc.abstractmethod
-    def chunk(
-        self,
-        document: Document,
-        embedder: IndexingEmbedder,
-    ) -> list[DocAwareChunk]:
-        raise NotImplementedError
 
+def get_default_chunker(embedder: IndexingEmbedder) -> Chunker:
```
Comment (rkuo-danswer) [inline backend/danswer/indexing/chunker.py]: Similar comment to above ... maybe have a single get_chunker method with optional params.  I think the word "default" in get_default_chunker seems irrelevant.

File: backend/danswer/indexing/chunker.py [hunk 4]
```diff
@@ -122,192 +72,247 @@ def _get_metadata_suffix_for_document_index(
...
+            metadata_suffix_semantic = ""
+
+        return self._chunk_document(
+            document, title_prefix, metadata_suffix_semantic, metadata_suffix_keyword
         )
 
-    # If the chunk does not have any useable content, it will not be indexed
-    return chunks
 
+_DEFAULT_CHUNKER: Chunker | None = None
```
Comment (rkuo-danswer) [inline backend/danswer/indexing/chunker.py]: is caching these chunkers globally a new optimization? I think these probably belong at the top of the file.  Could be a good optimization and would warrant.

File: backend/danswer/indexing/chunker.py [hunk 5]
```diff
@@ -122,192 +72,247 @@ def _get_metadata_suffix_for_document_index(
...
+                current_token_length
+                + len(self.tokenizer.tokenize(SECTION_SEPARATOR))
+                + section_token_length
+                <= self.chunk_token_limit
+            ):
+                chunk_text += SECTION_SEPARATOR * bool(chunk_text) + section_text
+                link_offsets[curr_offset_len] = section_link_text
+            else:
                 chunks.append(
                     DocAwareChunk(
```
Comment (rkuo-danswer) [inline backend/danswer/indexing/chunker.py]: The code for creating chunks seems to share several parameters, so perhaps a new Factory class that takes those parameters once and avoids us having to pass them around repeatedly could help.

The rest of the parameters seem to be quite different in various places, so it raises concerns that perhaps we've jammed too many parameters into DocAwareChunk.  It's possible the different usages are really different objects that need to be initialized differently but share a similar chunk interface.

File: backend/danswer/indexing/chunker.py [hunk 6]
```diff
@@ -122,192 +72,247 @@ def _get_metadata_suffix_for_document_index(
...
             )
+            metadata_tokens = len(self.tokenizer.tokenize(metadata_suffix_semantic))
+
+        if metadata_tokens >= self.chunk_token_limit * MAX_METADATA_PERCENTAGE:
+            # Note: we can keep the keyword suffix even if the semantic suffix is too long to fit in the model
+            # context, there is no limit for the keyword component
+            metadata_suffix_semantic = ""
+            metadata_tokens = 0
+
+        content_token_limit = self.chunk_token_limit - title_tokens - metadata_tokens
```
Comment (rkuo-danswer) [inline backend/danswer/indexing/chunker.py:285]: Where does content_token_limit get used?

**Review by rkuo-danswer — CHANGES_REQUESTED** [review]
Overall some useful changes here, especially since a lot of methods are now clearly private to the class and can be ignored by people reading the code.  Several comments inline.

File: backend/danswer/indexing/chunker.py (hunk 6 above)
Comment (hagen-danswer) [reply backend/danswer/indexing/chunker.py:285]: Good catch, i pass it into the chunk logic instead of using the class variable (because we dont want the actual chunking process to affect the state of the chunker)

File: backend/danswer/configs/app_configs.py (hunk 1 above)
Comment (hagen-danswer) [reply backend/danswer/configs/app_configs.py]: We may just define it as a 4-1 ratio if we are unable to figure out some whitespace counting discrepancies

File: backend/danswer/document_index/vespa_constants.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
+from danswer.configs.app_configs import VESPA_CONFIG_SERVER_HOST
```
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa_constants.py:1]: Moved constants that were only used with vespa into this file for better organization

File: backend/danswer/indexing/chunker.py (hunk 4 above)
Comment (hagen-danswer) [reply backend/danswer/indexing/chunker.py]: This is an optimization. The value of the optimization is unconfirmed. 

The other places we have implemented this has been below the class it is caching so im trying to keep it consistent

File: backend/danswer/indexing/chunker.py
Comment (hagen-danswer) [inline backend/danswer/indexing/chunker.py:1]: Here, we create a chunker object that defaults to using 512 size. but can be reused to chunk at a larger size as well

File: backend/danswer/indexing/embedder.py
Comment (hagen-danswer) [inline backend/danswer/indexing/embedder.py:1]: I think this is a good place to add time logging as it is time intensive

File: backend/danswer/indexing/indexing_pipeline.py [hunk 8]
```diff
@@ -181,12 +183,34 @@ def index_doc_batch(
     )
 
     logger.debug("Starting chunking")
```
Comment (hagen-danswer) [inline backend/danswer/indexing/indexing_pipeline.py:184]: this should be removed prior to merge

File: backend/danswer/indexing/indexing_pipeline.py [hunk 9]
```diff
@@ -181,12 +183,34 @@ def index_doc_batch(
...
-        for chunk in chunker.chunk(document=document, embedder=embedder)
-    ]
+    chunker: Chunker = get_default_chunker(embedder)
+    if ENABLE_MEGA_CHUNK:
+        mega_chunker: Chunker = get_mega_chunker(embedder)
+    chunks: list[DocAwareChunk] = []
+    for document in updatable_docs:
+        chunks_per_document = chunker.chunk(document=document)
+        chunks.extend(chunks_per_document)
+        if ENABLE_MEGA_CHUNK:
```
Comment (hagen-danswer) [inline backend/danswer/indexing/indexing_pipeline.py]: if mega chunks are enabled, we rechunk the document using the mega chunker, add references to the regular size chunks to the mega chunks, and add that to the final chunks list

File: backend/danswer/indexing/models.py
Comment (hagen-danswer) [inline backend/danswer/indexing/models.py:1]: mega chunks are defined by having references to smaller chunks

File: backend/danswer/search/pipeline.py [hunk 10]
```diff
@@ -166,49 +168,51 @@ def _get_sections(self) -> list[InferenceSection]:
...
         # Full doc setting takes priority
 
         if self.search_query.full_doc:
             seen_document_ids = set()
-            unique_chunks = []
 
             # This preserves the ordering since the chunks are retrieved in score order
             for chunk in retrieved_chunks:
                 if chunk.document_id not in seen_document_ids:
                     seen_document_ids.add(chunk.document_id)
```
Comment (hagen-danswer) [inline backend/danswer/search/pipeline.py:188]: parallelization is taken care of in the id_based_retrieval function itself now

File: backend/danswer/search/pipeline.py [hunk 11]
```diff
@@ -245,37 +249,34 @@ def _get_sections(self) -> list[InferenceSection]:
...
-        flattened_inference_chunks: list[InferenceChunk] = []
-        parallel_functions_with_args = []
 
         for chunk_range in flat_ranges:
             # Don't need to fetch chunks within range for merging if chunk_above / below are 0.
             if above == below == 0:
-                flattened_inference_chunks.extend(chunk_range.chunks)
+                inference_chunks.extend(chunk_range.chunks)
 
             else:
```
Comment (hagen-danswer) [inline backend/danswer/search/pipeline.py:257]: parallelization is taken care of in the id_based_retrieval function itself now

File: backend/danswer/search/retrieval/search_runner.py [hunk 12]
```diff
@@ -163,7 +174,62 @@ def doc_index_retrieval(
         else:
             raise RuntimeError("Invalid Search Flow")
 
-    return cleanup_chunks(top_chunks)
+    retrieval_requests: list[VespaChunkRequest] = []
```
Comment (hagen-danswer) [inline backend/danswer/search/retrieval/search_runner.py:151]: This part of the code removes the mega chunks, saves the score of each mega chunk, retrieves the original chunks referenced by each mega chunk, gives the score of the mega chunk to the normal chunks, dedupes the normal chunks (keeping the higher score) and finally reorders with descending score

File: backend/danswer/search/retrieval/search_runner.py [hunk 13]
```diff
@@ -249,34 +315,42 @@ def inference_sections_from_ids(
...
+    retrieved_chunks = document_index.id_based_retrieval(
+        chunk_requests=chunk_requests,
+        filters=filters,
     )
 
-    # Any failures to retrieve would give a None, drop the Nones and empty lists
-    inference_chunks_sets = [res for res in parallel_results if res]
+    cleaned_chunks = cleanup_chunks(retrieved_chunks)
+    if not cleaned_chunks:
+        return []
```
Comment (hagen-danswer) [inline backend/danswer/search/retrieval/search_runner.py:315]: parallelization is taken care of in the id_based_retrieval function itself now

File: backend/danswer/server/documents/document.py [hunk 14]
```diff
@@ -35,10 +37,8 @@ def get_document_info(
```
Comment (hagen-danswer) [inline backend/danswer/server/documents/document.py:37]: use new VespaChunkRequest object

File: backend/danswer/document_index/vespa/shared_utils/utils.py
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa/shared_utils/utils.py:1]: just moved this to be in the shared_utils folder

File: backend/danswer/document_index/vespa/shared_utils/vespa_request_builders.py
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa/shared_utils/vespa_request_builders.py:1]: This file is to hold the request string builders.

File: backend/danswer/document_index/vespa/indexing_utils.py
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa/indexing_utils.py:1]: These are the utils for index.py to actually perform the indexing

File: backend/danswer/document_index/vespa/deletion.py
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa/deletion.py:1]: these are the deletion utils

File: backend/danswer/document_index/vespa/chunk_retrieval.py
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa/chunk_retrieval.py:1]: These are the utils for chunk retrieval. I actually made some changes here!

File: backend/danswer/document_index/vespa/chunk_retrieval.py [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+        flattened_vespa_chunks.extend(chunk_set)
+
+    inference_chunks = [
+        _vespa_hit_to_inference_chunk(chunk, null_score=True)
+        for chunk in flattened_vespa_chunks
+    ]
+
+    return inference_chunks
+
+
```
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa/chunk_retrieval.py:279]: This is used to batch retrieve chunks which is much faster (10x) than using the visit api.

However, it only works if we know the number of chunks we are trying to get (otherwise, we may request too many. In the instances we dont know the range, we make the remaining calls using the visit api

File: backend/danswer/document_index/vespa/chunk_retrieval.py [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+    document_chunks = _get_chunks_via_visit_api(
+        chunk_request=VespaChunkRequest(document_id=document_id),
+        index_name=index_name,
+        filters=filters or IndexFilters(access_control_list=None),
+        field_names=[DOCUMENT_ID],
+        get_mega_chunks=get_mega_chunks,
+    )
+    return [chunk["id"].split("::", 1)[-1] for chunk in document_chunks]
+
+
```
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa/chunk_retrieval.py:246]: This brings the parallelization of the visit api calls here so we dont have to repeat the parallelization logic everywhere else in the code

File: backend/danswer/document_index/interfaces.py [hunk 17]
```diff
@@ -16,6 +16,13 @@ class DocumentInsertionRecord:
     already_existed: bool
```
Comment (hagen-danswer) [inline backend/danswer/document_index/interfaces.py:18]: The point of this new data class is to standardize chunk requests.
This may be combined with the other data type used for fetching chunks above and below in the future...

File: backend/danswer/configs/constants.py
Comment (hagen-danswer) [inline backend/danswer/configs/constants.py:1]: moved these into document_index folder

File: backend/danswer/document_index/vespa/index.py
Comment (hagen-danswer) [inline backend/danswer/document_index/vespa/index.py:1]: I moved a bunch of the utils that were in here into their own modules.

**Review by hagen-danswer — COMMENTED** [review]
I tried to leave comments explaining the changes for each module to help guide people through the PR. Let me know if anyone has anyquestions!

File: backend/danswer/document_index/vespa/deletion.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+            res = http_client.delete(
+                f"{DOCUMENT_ID_ENDPOINT.format(index_name=index_name)}/{chunk_id}"
+            )
+            res.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            logger.error(f"Failed to delete chunk, details: {e.response.text}")
+            raise
+
+
+def delete_vespa_docs(
```
Comment (rkuo-danswer) [inline backend/danswer/document_index/vespa/deletion.py:40]: Could apply the "Decorator" pattern to this.  Make a simpler helper function that always accepts an non-optional executor parameter. Then wrap it with another function that introduces the idea of an optional executor.

File: backend/danswer/search/retrieval/search_runner.py [hunk 19]
```diff
@@ -163,7 +174,62 @@ def doc_index_retrieval(
...
+            if key not in unique_chunks or (chunk.score or 0) > (
+                unique_chunks[key].score or 0
+            ):
+                unique_chunks[key] = chunk
+
+        deduped_chunks = list(unique_chunks.values())
+        deduped_chunks.sort(key=lambda chunk: chunk.score or 0, reverse=True)
+        return cleanup_chunks(deduped_chunks)
+
+    return cleanup_chunks(normal_chunks)
```
Comment (rkuo-danswer) [inline backend/danswer/search/retrieval/search_runner.py]: it would reduce the amount of nesting if you put this return first .. e.g. "if not retrieval_requests:"

File: backend/danswer/document_index/vespa/chunk_retrieval.py [hunk 20]
```diff
@@ -0,0 +1,418 @@
...
+    get_mega_chunks: bool = False,
+) -> list[InferenceChunkUncleaned]:
+    retrieved_chunks: list[InferenceChunkUncleaned] = []
+    capped_requests: list[VespaChunkRequest] = []
+    uncapped_requests: list[VespaChunkRequest] = []
+    chunk_count = 0
+    for request in chunk_requests:
+        # All requests without a chunk range are uncapped
+        # Uncapped requests are retrieved using the Visit API
+        if request.max_chunk_ind is None or request.min_chunk_ind is None:
```
Comment (rkuo-danswer) [inline backend/danswer/document_index/vespa/chunk_retrieval.py]: Could the calculations here be part of the VespaChunkRequest class?  Aka .is_uncapped(), etc?

File: backend/danswer/document_index/vespa/indexing_utils.py [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+from danswer.document_index.vespa_constants import TITLE
+from danswer.document_index.vespa_constants import TITLE_EMBEDDING
+from danswer.indexing.models import DocMetadataAwareIndexChunk
+from danswer.utils.logger import setup_logger
+
+logger = setup_logger()
+
+
+@retry(tries=3, delay=1, backoff=2)
+def _does_document_exist(
```
Comment (rkuo-danswer) [inline backend/danswer/document_index/vespa/indexing_utils.py:49]: feels like interactions like this with Vespa belong grouped together in some sort of class client library.  Doesnt have to be in this PR but maybe a TODO in linear.

**Review by rkuo-danswer — CHANGES_REQUESTED** [review]
Reviewed as much as I could ... it's a very large PR for which I don't have much context, but I did make some suggestions where I could.

File: backend/danswer/document_index/vespa/app_config/schemas/danswer_chunk.sd [hunk 22]
```diff
@@ -88,6 +88,9 @@ schema DANSWER_CHUNK_NAME {
             rank:filter
             attribute: fast-search
         }
```
Comment (yuhongsun96) [inline backend/danswer/document_index/vespa/app_config/schemas/danswer_chunk.sd:90]: add a comment on the Vespa field because it's not immediately clear to reader what this is

File: backend/danswer/natural_language_processing/search_nlp_models.py [hunk 23]
```diff
@@ -95,12 +97,19 @@ def __init__(
...
         self.model_name = model_name
         self.retrim_content = retrim_content
+        self.tokenizer = get_tokenizer(
+            model_name=model_name, provider_type=provider_type
+        )
+
+        self.max_seq_length = max_seq_length
+        search_settings = get_search_settings()
+        if search_settings and search_settings.multipass_indexing:
+            self.max_seq_length *= LARGE_CHUNK_RATIO
```
Comment (yuhongsun96) [inline backend/danswer/natural_language_processing/search_nlp_models.py]: can we add a comment to DOC_EMBEDDING_CONTEXT_SIZE to say this will be multiplied by the LARGE_CHUNK_RATIO if multipass is on?

Comment (yuhongsun96) [inline backend/danswer/natural_language_processing/search_nlp_models.py]: have some instance var which checks if this embedding model can even take large chunks. If not, skip it during the batch encode text based on if it has the pointers to the normal chunks.

Only allow if provider_type or is a nomic model (check the names)

Comment (yuhongsun96) [inline backend/danswer/natural_language_processing/search_nlp_models.py]: be sure to update the max_seq_length correctly so the model server doesnt throw exceptions

File: backend/danswer/document_index/document_index_utils.py [hunk 24]
```diff
@@ -50,4 +50,11 @@ def get_uuid_from_chunk(
     unique_identifier_string = "_".join(
         [doc_str, str(chunk.chunk_id), str(mini_chunk_ind)]
     )
+    if chunk.large_chunk_reference_ids:
+        unique_identifier_string += "_".join(
```
Comment (yuhongsun96) [inline backend/danswer/document_index/document_index_utils.py]: we can either prefix this with large or remove the mini chunk id

Comment (yuhongsun96) [reply backend/danswer/document_index/document_index_utils.py]: I think the mini chunk id is never used anyhow, should maybe just get rid of it

File: backend/danswer/document_index/interfaces.py [hunk 25]
```diff
@@ -16,6 +16,23 @@ class DocumentInsertionRecord:
...
 
 
+@dataclass(frozen=True)
+class VespaChunkRequest:
+    document_id: str
+    min_chunk_ind: int | None = None
+    max_chunk_ind: int | None = None
+
+    def is_capped(self) -> bool:
```
Comment (yuhongsun96) [inline backend/danswer/document_index/interfaces.py:26]: can we make these properties?

File: backend/danswer/document_index/interfaces.py [hunk 26]
```diff
@@ -16,6 +16,23 @@ class DocumentInsertionRecord:
...
+        # If the max chunk index is not None, then the chunk request is capped
+        # If the min chunk index is None, we can assume the min is 0
+        return self.max_chunk_ind is not None
+
+    def range(self) -> int | None:
```
Comment (yuhongsun96) [inline backend/danswer/document_index/interfaces.py:32]: would use a max(x, 0) in these cases.

File: backend/danswer/document_index/vespa/chunk_retrieval.py [hunk 27]
```diff
@@ -0,0 +1,425 @@
...
+
+    # parse fields that are stored as strings, but are really json / datetime
+    metadata = json.loads(fields[METADATA]) if METADATA in fields else {}
+    updated_at = (
+        datetime.fromtimestamp(fields[DOC_UPDATED_AT], tz=timezone.utc)
+        if DOC_UPDATED_AT in fields
+        else None
+    )
+
+    # The highlights might include the title but this is the best way we have so far to show the highlighting
```
Comment (yuhongsun96) [inline backend/danswer/document_index/vespa/chunk_retrieval.py]: This comment is no longer applicable, can just remove it

File: backend/tests/regression/answer_quality/file_uploader.py [hunk 28]
```diff
@@ -52,6 +52,7 @@ def upload_test_files(zip_file_path: str, env_name: str) -> None:
```
Comment (yuhongsun96) [inline backend/tests/regression/answer_quality/file_uploader.py:52]: Not here! but for the tests, can you mock the call to get search settings just to make the failing test pass?

**Review by yuhongsun96 — APPROVED** [review]
lgtm (didn't thoroughly read it)

**Review by yuhongsun96 — APPROVED** [review]
lgtm
